File: android/src/main/java/com/memotalk/sdk/RecordingManager.kt
package com.memotalk.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class RecordingManager(private val bleClient: BleClient) {

    suspend fun startRecording(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x80: Start recording
            // Parameters: status=1 (start), transfer_type (0=BLE, 1=WiFi)
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x80 // Command

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(4001, "Start recording failed: ${e.message}"))
        }
    }

    suspend fun stopRecording(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x80: Stop recording
            // Parameters: status=0 (stop)
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x80 // Command

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(4002, "Stop recording failed: ${e.message}"))
        }
    }

    suspend fun pauseRecording(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x80: Pause recording
            // Parameters: status=2 (pause)
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x80 // Command

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(4003, "Pause recording failed: ${e.message}"))
        }
    }

    suspend fun resumeRecording(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x80: Resume recording
            // Parameters: status=3 (resume)
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x80 // Command

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(4004, "Resume recording failed: ${e.message}"))
        }
    }

    suspend fun getRecordingList(device: Device): Result<List<Recording>, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x02: Get recording status
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x02 // Command

            bleClient.sendCommand(command)
            // Response parsing logic will be implemented
            Result.success(emptyList())
        } catch (e: Exception) {
            Result.failure(PlaudError(4005, "Get recording list failed: ${e.message}"))
        }
    }

    suspend fun clearAllRecordings(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x9A: Clear all recordings
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x9A // Command

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(4006, "Clear recordings failed: ${e.message}"))
        }
    }
}

File: ios/RecordingManager.swift
import Foundation

public class RecordingManager {

    private let bleClient: BleClient

    public init(bleClient: BleClient) {
        self.bleClient = bleClient
    }

    public func startRecording(device: Device) async -> Result<Void, PlaudError> {
        do {
            // Command 0x80: Start recording
            let command = Data([0x5A, 0x03, 0x80])
            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 4001, message: "Start recording failed: \(error.localizedDescription)"))
        }
    }

    public func stopRecording(device: Device) async -> Result<Void, PlaudError> {
        do {
            // Command 0x80: Stop recording
            let command = Data([0x5A, 0x03, 0x80])
            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 4002, message: "Stop recording failed: \(error.localizedDescription)"))
        }
    }

    public func pauseRecording(device: Device) async -> Result<Void, PlaudError> {
        do {
            // Command 0x80: Pause recording
            let command = Data([0x5A, 0x03, 0x80])
            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 4003, message: "Pause recording failed: \(error.localizedDescription)"))
        }
    }

    public func resumeRecording(device: Device) async -> Result<Void, PlaudError> {
        do {
            // Command 0x80: Resume recording
            let command = Data([0x5A, 0x03, 0x80])
            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 4004, message: "Resume recording failed: \(error.localizedDescription)"))
        }
    }

    public func getRecordingList(device: Device) async -> Result<[Recording], PlaudError> {
        do {
            // Command 0x02: Get recording status
            let command = Data([0x5A, 0x03, 0x02])
            bleClient.sendCommand(command: command)
            // Response parsing logic will be implemented
            return .success([])
        } catch {
            return .failure(PlaudError(code: 4005, message: "Get recording list failed: \(error.localizedDescription)"))
        }
    }

    public func clearAllRecordings(device: Device) async -> Result<Void, PlaudError> {
        do {
            // Command 0x9A: Clear all recordings
            let command = Data([0x5A, 0x03, 0x9A])
            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 4006, message: "Clear recordings failed: \(error.localizedDescription)"))
        }
    }
}
